core/src/main/java/org/testcontainers/images/TimeLimitedLoggedPullImageResultCallback.java

```diff
@@ -0,0 +1,117 @@
+package org.testcontainers.images;
+
+import com.github.dockerjava.api.model.PullResponseItem;
+import com.github.dockerjava.core.command.PullImageResultCallback;
+import org.slf4j.Logger;
+import org.testcontainers.utility.TestcontainersConfiguration;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.testcontainers.DockerClientFactory.TESTCONTAINERS_THREAD_GROUP;
+
+/**
+ * {@link PullImageResultCallback} with improved logging of pull progress and a 'watchdog' which will abort the pull
+ * if progress is not being made, to prevent a hanging test
+ */
+public class TimeLimitedLoggedPullImageResultCallback extends LoggedPullImageResultCallback {
```

Comment:
I just realized... maybe we should make it internal? We have a higher level abstraction (`RemoteDockerImage`), and less public classes is always better :)

Reply:
Good point, I thought I had that at some point but I must have regressed it.

Reply:
Fixed, in a way 😄 

It's now package-private. However, there was one cross-package usage, from `org.testcontainers.DockerClientFactory#checkAndPullImage`.

I'm not actually sure that `checkAndPullImage` really _should_ exist in the public API. I've:
* refactored it to use `RemoteDockerImage`, thus resolving the cross-package visibility and encapsulating pull behaviour a bit better
* marked `checkAndPullImage` as deprecated so that we can potentially remove it in the future.

WDYT?

Reply:
I think that's a good idea :+1: